private_transactions/src/lib.rs

```diff
@@ -0,0 +1,782 @@
+		where
+			T: Tracer,
+			V: VMTracer,
+	{
+		let mut env_info = self.client.env_info(block).ok_or(PrivateTransactionError::StatePruned)?;
+		env_info.gas_limit = transaction.gas;
+
+		let mut state = self.client.state_at(block).ok_or(PrivateTransactionError::StatePruned)?;
+		// TODO: in case of BlockId::Latest these need to operate on the same state
+		let contract_address = match &transaction.action {
+			&Action::Call(ref contract_address) => {
+				let contract_code = Arc::new(self.get_decrypted_code(contract_address, block)?);
+				let contract_state = self.get_decrypted_state(contract_address, block)?;
+				trace!("Patching contract at {:?}, code: {:?}, state: {:?}", contract_address, contract_code, contract_state);
+				state.patch_account(contract_address, contract_code.clone(), Self::snapshot_to_storage(contract_state))?;
+				Some(contract_address.clone())
+			},
+			&Action::Create => None,
+		};
+
+		let engine = self.client.engine();
+		let contract_address = contract_address.or({
+			let sender = transaction.sender();
+			let nonce = state.nonce(&sender)?;
+			let (new_address, _) = ethcore_contract_address(engine.create_address_scheme(env_info.number), &sender, &nonce, &transaction.data);
+			Some(new_address)
+		});
+		let result = Executive::new(&mut state, &env_info, engine.machine()).transact_virtual(transaction, options)?;
+		let (encrypted_code, encrypted_storage) = match contract_address {
+			Some(address) => {
+				let (code, storage) = state.into_account(&address)?;
+				let enc_code = match code {
+					Some(c) => Some(self.encrypt(&address, &Self::iv_from_address(&address), &c)?),
+					None => None,
+				};
+				(enc_code, self.encrypt(&address, &Self::iv_from_transaction(transaction), &Self::snapshot_from_storage(&storage))?)
+			},
+			None => return Err(PrivateTransactionError::ContractDoesNotExist.into())
+		};
+		trace!("Private contract executed. code: {:?}, state: {:?}, result: {:?}", encrypted_code, encrypted_storage, result.output);
+		Ok(PrivateExecutionResult {
+			code: encrypted_code,
+			state: encrypted_storage,
+			result: result,
+		})
+	}
+
+	fn generate_constructor(validators: &[Address], code: Bytes, storage: Bytes) -> Bytes {
+		let constructor_code = DEFAULT_STUB_CONTRACT.from_hex().expect("Default contract code is valid");
+		let constructor = ethabi::Constructor { inputs: vec![
+			ethabi::Param { name: "v".into(), kind: ethabi::ParamType::Array(Box::new(ethabi::ParamType::Address)) },
+			ethabi::Param { name: "c".into(), kind: ethabi::ParamType::Bytes },
+			ethabi::Param { name: "s".into(), kind: ethabi::ParamType::Bytes },
+		]};
+
+		let tokens = [
+			ethabi::Token::Array(validators.iter().map(|a| ethabi::Token::Address(a.clone())).collect()),
+			ethabi::Token::Bytes(code),
+			ethabi::Token::Bytes(storage),
+		];
+
+		constructor.encode_input(constructor_code, &tokens).expect("Input is always valid")
```

Comment:
it's better to use `private::PrivateContract` which is typesafe. Ebove function could be replaced with:

```rust
let private = private::PrivateContract::default();
private.constructor(code, validators)
```